### ISSUE TITLE
Kubelet stats query fails for running containers

### ISSUE THREAD

**Issue by mbforbes** [GoogleCloudPlatform/kubernetes#8291]
This is blocking #7580 (reboot test).

In running a test that reboots all of the nodes and then runs the e2e tests, we saw a failure with the influxdb / heapster monitoring test:

```
Monitoring
  verify monitoring pods and all cluster nodes are available on influxdb using heapster.
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:51
[BeforeEach] Monitoring
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:42
>>> testContext.KubeConfig: /Users/maxforbes/.kube/config
[It] verify monitoring pods and all cluster nodes are available on influxdb using heapster.
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:51
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
...
<lots more of these>
...
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative

• Failure [303.125 seconds]
Monitoring
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:52
  verify monitoring pods and all cluster nodes are available on influxdb using heapster. [It]
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:51

  monitoring using heapster and influxdb test failed

  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:236
```

The heapster logs, from running `kubectl log monitoring-heapster-controller-xxxxx heapster` ([output](https://gist.github.com/mbforbes/214b59e41d12c1e377cb)), shows a ton of the following kubelet errors for all containers (this one happens to be for skydns):

```
...
2015-05-14T21:43:07.429325471Z E0514 21:43:07.427620      10 kubelet.go:85] failed to get stats from kubelet url: http://10.240.117.245:10255/stats/default/kube-dns-i7v2g/14002c59-fa6e-11e4-a2fd-42010af0174f/skydns - request failed - "500 Internal Server Error", response: "Internal Error: unable to find Docker container \"8f489c214090c89d5c2773507b2505ab47cabe73eb82e5ac5c9c875aba4d6941\"\n"
2015-05-14T21:43:07.429353483Z E0514 21:43:07.427643      10 kube_pods.go:106] failed to get stats for container "skydns" in pod "default"/"kube-dns-i7v2g"
...
```

The [kubelet log on the node (massive)](https://gist.githubusercontent.com/mbforbes/64624f694de522f79047/raw/gistfile1.txt) shows a bunch of 500 errors, such as:

```
I0514 22:05:07.427063    2503 server.go:158] HTTP InternalServerError: Internal Error: unable to find Docker container "8f489c214090c89d5c2773507b2505ab47cabe73eb82e5ac5c9c875aba4d6941"
I0514 22:05:07.427148    2503 server.go:636] GET /stats/default/kube-dns-i7v2g/14002c59-fa6e-11e4-a2fd-42010af0174f/skydns: (23.441891ms) 500
goroutine 4144 [running]:
github.com/GoogleCloudPlatform/kubernetes/pkg/httplog.(*respLogger).WriteHeader(0xc208c9a9c0, 0x1f4)
	/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/httplog/log.go:188 +0x9a
net/http.Error(0x7eff7cbcf080, 0xc208c9a9c0, 0xc20a71d400, 0x72, 0x1f4)
	/usr/src/go/src/net/http/server.go:1274 +0x9a
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet.(*Server).error(0xc20827ca20, 0x7eff7cbcf080, 0xc208c9a9c0, 0x7eff7cd57d80, 0xc208f92fa0)
	/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/server.go:159 +0x213
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet.(*Server).serveStats(0xc20827ca20, 0x7eff7cbcf080, 0xc208c9a9c0, 0xc20a3709c0)
	/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/server.go:726 +0x78f
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet.(*Server).handleStats(0xc20827ca20, 0x7eff7cbcf080, 0xc208c9a9c0, 0xc20a3709c0)
	/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/server.go:329 +0x46
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet.*Server.(github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet.handleStats)Â·fm(0x7eff7cbcf080, 0xc208c9a9c0, 0xc20a3709c0)
	/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/server.go:85 +0x45
net/http.HandlerFunc.ServeHTTP(0xc20833cd40, 0x7eff7cbcf080, 0xc208c9a9c0, 0xc20a3709c0)
	/usr/src/go/src/net/http/server.go:1265 +0x41
net/http.(*ServeMux).ServeHTTP(0xc2084094d0, 0x7eff7cbcf080, 0xc208c9a9c0, 0xc20a3709c0)
	/usr/src/go/src/net/http/server.go:1541 +0x17d
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet.(*Server).ServeHTTP(0xc20827ca20, 0x7eff7cd876f8, 0xc20ad308c0, 0xc20a3709c0)
	/go/src/github.com/GoogleCloudPlatform/kubernetes/_out [[Go 1.1 package http] 10.245.0.5:35262]
```

However,

- the apiserver thinks that all the pods are running (`kubectl get pods` gives this [output](https://gist.github.com/mbforbes/0aded988fa727fa3cab8))
- Docker thinks that the pods are running (`sudo docker ps` on that node gives this [output](https://gist.github.com/mbforbes/e860bf28a399e4a2dbdf))

In summary:
- apiserver and docker think containers are running
- but kubelet can't find them
- this manifested in the influxdb / heapster monitoring e2e test after node reboot

I've still got the cluster around if anyone would like to debug.

+cc @vishh (helped me debug this (thanks!))
+cc @dchen1107 (node)
+cc @roberthbailey (interested in getting #7580 in, which this is blocking)

**Comment by dchen1107** [comment]
@mbforbes Can I access your test nodes?

**Comment by vmarmol** [comment]
Is this an active problem? I'd check if the cAdvisor thinks the containers exist (it is the one returning the error). I'd then scan the logs to see if cAdvisor had any errors related to that container.

**Comment by dchen1107** [comment]
@mbforbes Unfortunately during debugging, I restart kubelet service on that node, the issue is gone. The problem could be cadvisor cannot get stats, but I couldn't find any cadvisor related error messages. How easy to reproduce the issue here? Next time we might access cadvisor's page to see if there is no such container information. Thanks!

**Comment by mbforbes** [comment]
@dchen1107 (and @vmarmol) Thanks for looking into this! I believe I can reproduce easily but I'll confirm. Where do I look for cAdvisor info? I don't see a log in `/var/log/`or `/var/log/containers/`, and I don't see a cAdvisor container running?

**Comment by dchen1107** [comment]
cAdvisor is running as part of kubelet now. If you can reproduce the issue, could you please change kubelet's log level to 4 so that we have more cadvisor log here? Thanks!

**Comment by mbforbes** [comment]
@dchen1107 Yes! How do I do that?

**Comment by mbforbes** [comment]
I can reproduce this by doing the following:

```bash
go run hack/e2e.go --v --up

# The following uses #7580; equivalent to just running `sudo reboot` on each node.
go run hack/e2e.go --v --test --test_args="--ginkgo.focus=Reboot"  

# Run the rest of the tests and you can see the issue.
go run hack/e2e.go --v --test
```

After this, I got two failures:

```
[Fail] Services [AfterEach] should serve multiport endpoints from pods
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:284

[Fail] Monitoring [It] verify monitoring pods and all cluster nodes are available on influxdb using heapster.
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:236
```

... and I have another node with more of the same kubelet errors:

```
0515 03:37:06.205414    2502 server.go:158] HTTP InternalServerError: Internal Error: unable to find Docker container "76c1fb26e89a8ff17b36eabaa3bdfa0b993a41eea06425571893ccc9ee25fe74"
I0515 03:37:06.205492    2502 server.go:636] GET /stats/default/kube-dns-eeu83/48f434b1-faa4-11e4-b7f3-42010af021f7/kube2sky: (14.396381ms) 500
goroutine 3075 [running]:
github.com/GoogleCloudPlatform/kubernetes/pkg/httplog.(*respLogger).WriteHeader(0xc208a4c660, 0x1f4)
	/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/httplog/log.go:188 +0x9a
net/http.Error(0x7f2259b187a8, 0xc208a4c660, 0xc20951c200, 0x72, 0x1f4)
	/usr/src/go/src/net/http/server.go:1274 +0x9a
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet.(*Server).error(0xc20848a1c0, 0x7f2259b187a8, 0xc208a4c660, 0x7f2259ca2d80, 0xc209668660)
	/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/server.go:159 +0x213
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet.(*Server).serveStats(0xc20848a1c0, 0x7f2259b187a8, 0xc208a4c660, 0xc2095616c0)
	/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/server.go:726 +0x78f
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet.(*Server).handleStats(0xc20848a1c0, 0x7f2259b187a8, 0xc208a4c660, 0xc2095616c0)
	/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/server.go:329 +0x46
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet.*Server.(github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet.handleStats)·fm(0x7f2259b187a8, 0xc208a4c660, 0xc2095616c0)
	/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/server.go:85 +0x45
net/http.HandlerFunc.ServeHTTP(0xc208527f00, 0x7f2259b187a8, 0xc208a4c660, 0xc2095616c0)
	/usr/src/go/src/net/http/server.go:1265 +0x41
net/http.(*ServeMux).ServeHTTP(0xc208525a40, 0x7f2259b187a8, 0xc208a4c660, 0xc2095616c0)
	/usr/src/go/src/net/http/server.go:1541 +0x17d
github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet.(*Server).ServeHTTP(0xc20848a1c0, 0x7f2259ccce50, 0xc2082d4000, 0xc2095616c0)
	/go/src/github.com/GoogleCloudPlatform/kubernetes/_out [[Go 1.1 package http] 10.245.1.3:44921]
```

(If I need to change the Kubelet log level at compile-time I can do that and repro on a new cluster; otherwise, I've got another cluster with the issue that can be introspected...)

**Comment by dchen1107** [comment]
There is a race between cadvisor initialization and docker. cadvisor failed to register docker factory, and bailed out. @vmarmol is going to make cadvisor more resilient to the issue of docker is not available, and have some retry logic. I am going to change kubelet initialization logic a little bit to make sure docker is running before we start cadvisor.

**Comment by mbforbes** [comment]
Awesome, thank you Dawn! Are there issues or PRs I can follow to know when this is unblocked?

**Comment by vmarmol** [comment]
@mbforbes #8324 is one of them. @dchen1107 is working on the second.
